polyval/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+
+extern crate byteorder;
+extern crate crypto_mac;
+#[cfg(feature = "zeroize")]
+extern crate zeroize;
+
+use byteorder::{BigEndian, ByteOrder, LittleEndian};
+use core::cmp::min;
+use core::marker::PhantomData;
+use crypto_mac::generic_array::{typenum::U16, GenericArray};
+use crypto_mac::{Mac, MacResult};
+#[cfg(feature = "zeroize")]
+use zeroize::Zeroize;
+
+/// The POLYVAL universal hash function
+pub type Polyval = PolyvalCore<LittleEndian>;
+
+/// The GHASH universal hash function
+pub type GHash = PolyvalCore<BigEndian>;
+
+/// POLYVAL: GHASH-like universal hash over GF(2^128).
+///
+/// From RFC 8452, Appendix A:
+///
+/// > The Relationship between POLYVAL and GHASH
+/// >
+/// > GHASH and POLYVAL both operate in GF(2^128), although with different
+/// > irreducible polynomials: POLYVAL works modulo x^128 + x^127 + x^126 +
+/// > x^121 + 1 and GHASH works modulo x^128 + x^7 + x^2 + x + 1.  Note
+/// > that these irreducible polynomials are the "reverse" of each other.
+///
+/// Using this property, we are able to implement both the POLYVAL and GHASH
+/// functions in terms of POLYVAL simply by being generic over the byte order.
+/// However, POLYVAL has the advantage of being optimized for natively
+/// little-endian CPUs.
+#[repr(align(16))]
+#[derive(Clone)]
+pub struct PolyvalCore<B: ByteOrder> {
+    h: u128,
+    s: u128,
+    leftover: usize,
+    buffer: [u8; 16],
+    o: PhantomData<B>,
+}
+
+const XMM_MASK: u128 = 0x0001_c200_0000_0000_0000;
+
+// TODO(tarcieri): POLYVAL is a universal hash function, not a MAC.
+// It can be used to implement a MAC (e.g. GMAC), but should probably use
+// a different trait than `crypto_mac::Mac`.
+impl<B> Mac for PolyvalCore<B>
+where
+    B: ByteOrder,
+{
+    type KeySize = U16;
+    type OutputSize = U16;
+
+    fn new(h: &GenericArray<u8, Self::KeySize>) -> Self {
+        Self {
+            h: B::read_u128(h),
+            s: 0,
+            leftover: 0,
+            buffer: [0u8; 16],
+            o: PhantomData,
+        }
+    }
+
+    #[inline]
+    fn input(&mut self, data: &[u8]) {
+        let mut m = data;
+
+        if self.leftover > 0 {
+            let want = min(16 - self.leftover, m.len());
+
+            for (i, byte) in m.iter().cloned().enumerate().take(want) {
+                self.buffer[self.leftover + i] = byte;
+            }
+
+            m = &m[want..];
+            self.leftover += want;
+
+            if self.leftover < 16 {
+                return;
+            }
+
+            let tmp = self.buffer;
+            self.block(&tmp);
+            self.leftover = 0;
+        }
+
+        while m.len() >= 16 {
+            self.block(&m[..16]);
+            m = &m[16..];
+        }
+
+        self.buffer[..m.len()].copy_from_slice(m);
+        self.leftover = m.len();
+    }
+
+    fn result(mut self) -> MacResult<Self::OutputSize> {
+        // TODO(tarcieri): this is almost certainly wrong
```

Comment:
TODO: The code below is copypasta from the Poly1305 PR (#12), and does not properly reflect what should be done in this case.